### PR TITLE
SNOW-1617634: Update large query breakdown to not break and duplicate CTEs

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1617634

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe.

3. Please describe how your code solves the related issue.

   The PR makes the following changes:
  i. Update `referenced_ctes` to hold map of `WithQueryBlock` -> count of number of nodes referring to that `WithQueryBlock` in the subtree.
  ii. Add `WithQueryBlock` to an eligible pipeline breaker node
  iii. Update node eligibility criteria so that we make any node which has an external `WithQueryBlock` reference ineligible.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 1]
```diff
@@ -270,23 +293,75 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
...
+        reason_if_invalid = None
+        if not (
             self.complexity_score_lower_bound
             < score
             < self.complexity_score_upper_bound
-        ) and self._is_node_pipeline_breaker(node)
-        if valid_node:
+        ):
+            is_valid = False
+            reason_if_invalid = CompilationStageTelemetryField.INVALID_DUE_TO_SCORE
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:316]: let's get enum class about all reasons, and we can we can have more complete reason, for this one, we can even separate them to SCORE_LOW,  and SCORE_HIGH

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 2]
```diff
@@ -270,23 +293,75 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
...
+
+        if is_valid and not self._is_node_pipeline_breaker(node):
+            is_valid = False
+            reason_if_invalid = CompilationStageTelemetryField.INVALID_DUE_TO_PIPELINE
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:320]: NOT_PIPELINE_BREAKER

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 3]
```diff
@@ -309,6 +384,10 @@ def _is_node_pipeline_breaker(self, node: LogicalPlan) -> bool:
             # Except and Intersect are pipeline breakers since they are join + distinct
             return True
 
+        if isinstance(node, WithQueryBlock):
+            # WithQueryBlock is a pipeline breaker
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:388]: if we are treating the WithQueryBlock as a pipeline breaker, when we are breaking the WtihQueryBlock, we can probably directly remove the WiWtihQueryBlock, but propogate the child up, maybe we can get a ticket to do the improvement in a different pr

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 4]
```diff
@@ -124,6 +126,10 @@ def __init__(
         self.complexity_score_upper_bound = (
             session.large_query_breakdown_complexity_bounds[1]
         )
+        # This is used to track the number of partitions we could not breakdown because we
+        # could not find any valid nodes. We also track if we could have broken down the plan
+        # only if externally referenced CTEs were considered valid.
+        self.breakdown_failure_summary = defaultdict(int)
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:132]: instead of add it as a field, maybe do it similar as how we do it for the CTE reference, just return a structure result RepeatedSubqueryEliminationResult with the result summary. Or update the other one to do the same thing, so that we have a uniformed way of handling the optimizations

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 5]
```diff
@@ -233,9 +242,23 @@ def _find_node_to_breakdown(self, root: TreeNode) -> Optional[TreeNode]:
...
+                        if (
+                            reason_if_invalid
+                            == CompilationStageTelemetryField.INVALID_DUE_TO_EXTERNAL_CTE_REF
+                        ):
+                            invalid_due_to_external_cte_ref_encountered = True
 
             current_level = next_level
 
+        if candidate_node is None:
+            self.breakdown_failure_summary[
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:255]: instead of one summary, maybe we can get a list of summary for each logical plan each one includes 1) num_partitions_made 2) invalid nodes with reason
with those should be good enough for us to debug

File: tests/integ/test_large_query_breakdown.py [hunk 6]
```diff
@@ -112,6 +112,46 @@ def test_no_valid_nodes_found(session, caplog):
     assert "Could not find a valid node for partitioning" in caplog.text
 
 
+def test_large_query_breakdown_external_cte_ref(session):
```
Comment (sfc-gh-yzou) [inline tests/integ/test_large_query_breakdown.py:115]: do we have a test to test breakdown at cte node?

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py (hunk 3 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:431]: https://snowflakecomputing.atlassian.net/browse/SNOW-1734385

File: tests/integ/test_large_query_breakdown.py (hunk 6 above)
Comment (sfc-gh-aalam) [reply tests/integ/test_large_query_breakdown.py:115]: added now

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py (hunk 5 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py]: updated

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py (hunk 4 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py]: done

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 7]
```diff
@@ -186,23 +207,41 @@ def _try_to_breakdown_plan(self, root: TreeNode) -> List[LogicalPlan]:
...
-        # TODO: SNOW-1617634 Have a one pass algorithm to find the valid node for partitioning
+        final_partition_breakdown_summary = {}
         while complexity_score > self.complexity_score_upper_bound:
-            child = self._find_node_to_breakdown(root)
+            child, validity_statistics = self._find_node_to_breakdown(root)
             if child is None:
+                final_partition_breakdown_summary = {
+                    k.value: validity_statistics.get(k, 0)
+                    for k in [
+                        CompilationStageTelemetryField.INVALID_DUE_TO_SCORE_ABOVE_UPPER_BOUND,
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:217]: Why not have a enum class specific for InValidBreadownNodeReason so that we can group the reasons together, and then you don't need those "INVALID_DUE_TO" prefix, which could also make the code easier to understand.

File: tests/integ/test_large_query_breakdown.py [hunk 8]
```diff
@@ -116,6 +116,77 @@ def test_no_valid_nodes_found(session, caplog):
...
+        queries = final_df.queries
+
+    # assert that we did not break the plan
+    assert len(queries["queries"]) == 1
+
+    patch_send.assert_called_once()
+    _, kwargs = patch_send.call_args
+    summary_value = kwargs["compilation_stage_summary"]
+    assert summary_value["breakdown_failure_summary"] == [
+        {
```
Comment (sfc-gh-yzou) [inline tests/integ/test_large_query_breakdown.py:154]: can we add a mocked telemetry test for summary for more than one plan to be more clear on the format

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py (hunk 7 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:217]: Also if you have an enum class, we can simply iterate over the enum class by doing
" for reason in InValidBreadownNodeReason"

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 9]
```diff
@@ -382,7 +382,7 @@ def column_states(self, value: ColumnStateDict):
 
     @property
     @abstractmethod
-    def referenced_ctes(self) -> Set[WithQueryBlock]:
+    def referenced_ctes(self) -> Dict[WithQueryBlock, int]:
         """Return the set of ctes referenced by the whole selectable subtree, includes its-self and children"""
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:386]: we will need to update the comments for referenced_ctes to indicate the changes, where it also records the number of times each CTE is referred

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 10]
```diff
@@ -670,7 +669,11 @@ def build_binary(
             # duplicated queries if there is a common CTE block referenced by
             # both left and right.
             referenced_ctes.update(select_left.referenced_ctes)
-            referenced_ctes.update(select_right.referenced_ctes)
+            for with_query_block, count in select_right.referenced_ctes.items():
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:672]: I see a lot of if else, i think you can add a utility function in the query_plan_analysis_utils.py to help merge the refered_ctes which takes two dict.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 11]
```diff
@@ -1670,7 +1673,11 @@ def with_query_block(
         # query generation stage.
         queries = child.queries[:-1] + [Query(sql=new_query)]
         # propagate the cte table
-        referenced_ctes = {with_query_block}.union(child.referenced_ctes)
+        referenced_ctes = child.referenced_ctes.copy()
+        if with_query_block in referenced_ctes:
+            referenced_ctes[with_query_block] += 1
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:1678]: once you have the util funciton, you can just call it with {with_query_block: 1}

File: tests/integ/test_large_query_breakdown.py (hunk 8 above)
Comment (sfc-gh-aalam) [reply tests/integ/test_large_query_breakdown.py:154]: I added it for another test below. for CTE where the breakdown is successful

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 12]
```diff
@@ -267,23 +315,93 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
...
+        externally by node3.
+        Similarly, node3 contains a WithQueryBlock node that is also referenced externally
+        by node2.
+        However, node1 contains WithQueryBlock node that is not referenced externally.
+
+        We determine if a node contains an externally referenced CTE by comparing the
+        number of times each unique WithQueryBlock node is referenced in the subtree compared
+        to the number of times it is referenced in the root node.
+        """
+        if isinstance(node, SnowflakePlan) and isinstance(
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:391]: I will add explanation on why this is needed. Short version is the `SnowflakePlan` and `SelectSnowflake` plan automatically get marked as external cte ref without this special handling since their reference count for the `WithQueryBlock` they contain will always be 1. This will almost definitely be < root's count.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 9 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py]: fixed

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 13]
```diff
@@ -186,42 +209,64 @@ def _try_to_breakdown_plan(self, root: TreeNode) -> List[LogicalPlan]:
...
         plans = []
-        # TODO: SNOW-1617634 Have a one pass algorithm to find the valid node for partitioning
+        final_partition_breakdown_summary = {}
         while complexity_score > self.complexity_score_upper_bound:
-            child = self._find_node_to_breakdown(root)
+            child, validity_statistics = self._find_node_to_breakdown(root)
             if child is None:
+                final_partition_breakdown_summary = {
+                    k.value: validity_statistics.get(k, 0)
+                    for k in [
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:218]: you should be able to iterate over all enum values using 
```
for k in InvalidNodesInBreakdownCategory:
```

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 14]
```diff
@@ -267,31 +315,106 @@ def _get_partitioned_plan(self, root: TreeNode, child: TreeNode) -> SnowflakePla
...
+          SnowflakePlan         1         False
+
+        We determine if a node contains an externally referenced CTE by comparing the
+        number of times each unique WithQueryBlock node is referenced in the subtree compared
+        to the number of times it is referenced in the root node.
+        """
+
+        # Checks for SnowflakePlan and SelectSnowflakePlan is to prevent marking a WithQueryBlock, which is a pipeline breaker
+        # node as an external CTE ref.
+        if isinstance(node, SelectSnowflakePlan):
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:396]: here we need a special handling is because we want to apply the check on the snowflakePlan to handle the cte case mentioned below

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 15]
```diff
@@ -132,7 +155,7 @@ def apply(self) -> List[LogicalPlan]:
                 self.session.session_id,
                 SkipLargeQueryBreakdownCategory.ACTIVE_TRANSACTION.value,
             )
-            return self.logical_plans
+            return LargeQueryBreakdownResult(self.logical_plans, [])
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:158]: i see for the SkipLargeQueryBreakdown, it is sent as a separate telemetry, i think we can merge those, you can return {SkipLargeQueryBreakdownCategory.ACTIVE_TRANSACTION.value, 0} as s summary here, then we just need on telemetry at one place

**Review by sfc-gh-yzou — APPROVED** [review]

**Review by sfc-gh-jdu — APPROVED** [review]
